### PR TITLE
support phi3

### PR DESCRIPTION
**Pull request by grimoire**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/model.py [hunk 1]
```diff
@@ -1324,6 +1324,45 @@ def match(cls, model_path: str) -> Optional[str]:
...
+    """Chat template of InternLM model."""
+
+    def __init__(
+            self,
+            system='<|system|>\n',
+            meta_instruction="""You are a helpful AI assistant.""",  # noqa: E501
+            eosys='<|end|>\n',
+            user='<|user|>\n',
+            eoh='<|end|>\n',
+            assistant='<|assistant|>:',
```
Comment (RunningLeon) [inline lmdeploy/model.py:1338]: Shall we change `:` to `\n` according to the [config](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/blob/8a362e755d2faf8cec2bf98850ce2216023d178a/tokenizer_config.json#L119)?
```
"chat_template": "{{ bos_token }}{% for message in messages %}{% if (message['role'] == 'user') %}{{'<|user|>' + '\n' + message['content'] + '<|end|>' + '\n' + '<|assistant|>' + '\n'}}{% elif (message['role'] == 'assistant') %}{{message['content'] + '<|end|>' + '\n'}}{% endif %}{% endfor %}",
 ```

File: lmdeploy/model.py [hunk 2]
```diff
@@ -1324,6 +1324,45 @@ def match(cls, model_path: str) -> Optional[str]:
...
 
 
+@MODELS.register_module(name='phi-3')
+class Phi3Instruct(BaseChatTemplate):
+    """Chat template of InternLM model."""
+
+    def __init__(
+            self,
+            system='<|system|>\n',
+            meta_instruction="""You are a helpful AI assistant.""",  # noqa: E501
```
Comment (RunningLeon) [inline lmdeploy/model.py:1334]: where is this meta_instruction from?

File: lmdeploy/pytorch/models/phi3.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+                context._sin = sin
+            cos = context._cos
+            sin = context._sin
+            query_states, key_states = apply_rotary_pos_emb(
+                query_states,
+                key_states,
+                cos[0],
+                sin[0],
+                position_ids,
+                torch.arange(0,
```
Comment (RunningLeon) [inline lmdeploy/pytorch/models/phi3.py:98]: the `position_ids_1d` seems different. why not use `position_ids_1d` directly?

File: lmdeploy/model.py (hunk 2 above)
Comment (grimoire) [reply lmdeploy/model.py:1334]: https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/commit/fc313b0514906a348a624cc9421e36f52c60e137

File: lmdeploy/pytorch/models/phi3.py (hunk 3 above)
Comment (grimoire) [reply lmdeploy/pytorch/models/phi3.py:98]: https://github.com/InternLM/lmdeploy/blob/5852aaa0617aefb72099f3b1405daa9ff7035382/lmdeploy/pytorch/models/phi3.py#L86

**Review by RunningLeon — APPROVED** [review]
LGTM

File: lmdeploy/model.py [hunk 4]
```diff
@@ -1324,6 +1324,38 @@ def match(cls, model_path: str) -> Optional[str]:
...
+class Phi3Instruct(BaseChatTemplate):
+    """Chat template of InternLM model."""
+
+    def __init__(self,
+                 user='<|user|>\n',
+                 eoh='<|end|>\n',
+                 assistant='<|assistant|>\n',
+                 eoa='<|end|>\n',
+                 separator='\n',
+                 stop_words=['<|end|>'],
```
Comment (AllentDan) [inline lmdeploy/model.py:1337]: may add system instructions for possible user messages
```python
system='<|system|>\n'
meta_instruction=None
eosys='<|end|>\n'
```

Comment (AllentDan) [inline lmdeploy/model.py:1337]: `<|endoftext|>` seems also a stop word

**Review by AllentDan — APPROVED** [review]
Tested OK with `lmdeploy chat torch`

**Comment by lvhan028** [comment]
UT failed

**Pull request merged**